crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -0,0 +1,380 @@
+        vector_init(&ret->s1, params->l);
+    }
+    return ret;
+}
+
+/**
+ * @brief Destroy a ML_DSA_KEY object
+ */
+void ossl_ml_dsa_key_free(ML_DSA_KEY *key)
+{
+    int i;
+
+    if (key == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&key->references, &i);
+    REF_PRINT_COUNT("ML_DSA_KEY", key);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_free(key->pub_encoding);
+    OPENSSL_free(key->priv_encoding);
+    OPENSSL_free(key->propq);
+    CRYPTO_FREE_REF(&key->references);
+    OPENSSL_free(key);
+}
+
+/*
+ * @brief Increase the reference count for a ML_DSA_KEY object.
+ * @returns 1 on success or 0 otherwise.
+ */
+int ossl_ml_dsa_key_up_ref(ML_DSA_KEY *key)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&key->references, &i) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("ML_DSA_KEY", key);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+/**
+ * @brief Are 2 keys equal?
+ *
+ * To be equal the keys must have the same key data and algorithm name.
+ *
+ * @param key1 A ML_DSA_KEY object
+ * @param key2 A ML_DSA_KEY object
+ * @param selection to select public and/or private component comparison.
+ * @returns 1 if the keys are equal otherwise it returns 0.
+ */
+int ossl_ml_dsa_key_equal(const ML_DSA_KEY *key1, const ML_DSA_KEY *key2,
+                          int selection)
+{
+    if (key1->params != key2->params)
+        return 0;
+    if (key1->pub_encoding != NULL) {
+        if (key2->pub_encoding == NULL
+                || memcmp(key1->pub_encoding, key1->pub_encoding,
+                          key1->params->pk_len) != 0)
+            return 0;
+    } else if (key2->pub_encoding != NULL) {
+        return 0;
+    }
+    if (key1->priv_encoding != NULL) {
+        if (key2->priv_encoding == NULL
+                || memcmp(key1->priv_encoding, key1->priv_encoding,
+                          key1->params->sk_len) != 0)
+            return 0;
+    } else if (key2->priv_encoding != NULL) {
+        return 0;
+    }
+    return 1;
+}
+
+int ossl_ml_dsa_key_has(const ML_DSA_KEY *key, int selection)
+{
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        if (key->pub_encoding == NULL)
+            return 0; /* No public key */
+        if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0
+                && key->priv_encoding == 0)
+            return 0; /* No private key */
+        return 1;
+    }
+    return 0;
+}
+
+/**
+ * @brief Load a ML_DSA key from raw data.
+ *
+ * @param key An ML_DSA key to load into
+ * @param params An array of parameters containing key data.
+ * @param include_private Set to 1 to optionally include the private key data
+ *                        if it exists.
+ * @returns 1 on success, or 0 on failure.
+ */
+int ossl_ml_dsa_key_fromdata(ML_DSA_KEY *key, const OSSL_PARAM params[],
+                             int include_private)
+{
+    const OSSL_PARAM *p = NULL;
+
+    /* Private key is optional */
```

Comment:
This is odd to say, because the public key is likewise optional when `OSSL_PKEY_PARAM_PUB_KEY` was not specified.
Rather, the private key is in some sense *doubly optional*, because it will ignored, even when specified, if `include_private` is false.

Reply:
That is the pattern we use in other keys